staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,998 @@
+}
+
+// Valid [DeviceRequest.CountMode] values.
+const (
+	CountModeExact = "Exact"
+	CountModeAll   = "All"
+)
+
+// Exactly one field must be set.
+type Selector struct {
+	// CEL contains a CEL expression for selecting a device.
+	//
+	// +required
+	CEL *CELSelector `json:"cel,omitempty" protobuf:"bytes,1,opt,name=cel"`
+}
+
+// CELSelector contains a CEL expression for selecting a device.
+type CELSelector struct {
+	// This CEL expression must evaluate to true if a device is suitable.
+	// This covers qualitative aspects of device selection.
+	//
+	// The language is as defined in
+	// https://kubernetes.io/docs/reference/using-api/cel/
+	// with several additions that are specific to device selectors.
+	//
+	// Attributes of a device are made available through a nested
+	// `device.attributes` map with the domain part of the attribute name
+	// as key in the outer map and the identifier as key in the inner
+	// map. All identifiers can be used in a field lookup:
+	//
+	//    device.attributes["dra.example.com"].driverVersion
+	//
+	// The type of each entry varies, depending on the attribute
+	// that is being looked up. The domain lookup returns an empty
+	// map if there is no attribute with that domain. However,
+	// unknown identifiers then trigger a runtime error.
+	//
+	// The `cel.bind` function is enabled and can be used to simplify
+	// expressions that access multiple attributes with the same domain:
+	//
+	//    cel.bind(dra, device.attributes["dra.example.com"], dra.someBool && dra.anotherBool)
+	//
+	// Capacities associated with a device are made available through a
+	// nested `device.capacities` map the same way as attributes.
+	//
+	// The `device.driverName` string variable can be used to check for a specific
+	// driver explicitly in a filter that is meant to work for devices from
+	// different vendors. It is provided by Kubernetes and matches the
+	// `driverName` from the ResourceSlice which provides the device.
+	//
+	// The CEL expression is applied to *all* available devices from any driver.
+	// The expression has to check for existence of an attribute when it is not
+	// certain that it is provided because runtime errors are not automatically
+	// treated as "don't select device". Instead, device selection fails completely
+	// and reports the error.
+	//
+	// Some more examples:
+	//
+	//    "memory" in device.capacities["dra.example.com"] && # Is the capacity available?
+	//       device.capacities["dra.example.com"].memory.isGreaterThan(quantity("1Gi")) # >= 1Gi
+	//
+	//    device.attributes["dra.example.com"].driverVersion.isGreaterThan(semver("1.0.0")) # >= v1.0.0, runtime error if not available
+	//
+	//    device.driverName == "dra.example.com" # any device from that driver
+	//
+	// +required
+	Expression string `json:"expression" protobuf:"bytes,1,name=expression"`
+}
+
+// Besides the request name slice, constraint must have exactly one field set.
+type Constraint struct {
```

Comment:
Potential problem:

Requests[] is a list of requests, some of which are "devices", and some of which might be gizmos or widgets.  All of the selection criteria, in particular the__ whole idea__ of attribute evaluations, is subordinate to "device".

But constraints are a peer to requests.

Do constraints only apply to devices?  Or are we confident that gizmos and widgets will be structurally similar to devices (e.g. they will all have the notion of attributes which can be matched with devices)?  Or is it that constraints can apply across the set of devices but never between devices and gizmos?

What I am getting at is whether we need to say:

```
requests:
- device:
  countMode: Exact
  count: 4
  selectors:
  - cel:
    expression: device.driver == example.com && device.attributes["example.com"].genre == GPU
- gizmo:
  color: purple
  count: 4
constraints:
  device:
  - matchAttribute: model
  - matchAttribute: color
  gizmo:
  - sameShape: true 
 ```

or something even more dramatic?

Reply:
I agree that we need the `device` indirection also for constraints, same as for requests.

Reply:
Or rather, as for "selectors": we have a list of constraints and each entry is a single constraint of a particular type.

I changed it to:
```yaml
constraints:
- device:
    matchAttribute: model
```

Or perhaps this?
```yaml
constraints:
- matchDeviceAttribute: model
```

That's not that different from what we had above, except that it's clearer that this particular constraint applies to "devices".

Reply:
I do not love sprinkling "device" around ito each sub-struct.

Distilling.  Today we have:

```
type ResourceClaimSpec struct {
	Requests []Request
	Constraints []Constraint
	Config []ClaimConfiguration
}

type Request struct {
	Name string
	Device *DeviceRequest
}

type DeviceRequest struct {
	DeviceClassName string
	Selectors []Selector
	CountMode CountMode
	Count int64
}

type Selector struct {
	CEL *CELSelector
}

type Constraint struct {
	RequestNames []string
	MatchAttribute *string
}

type ClaimConfiguration struct {
	RequestNames []string
	Configuration
}
```

Requests and constraints and configuration are all peers and all (I think?)
need to distinguish devices from gizmos.  What if we moved devices to the top
of the hierarchy?

```
type ResourceClaimSpec struct {
	Devices []DeviceClaim
	// Gizmos []GizmoClaim
}

type DeviceClaim struct {
	Requests []Request
	Constraints []Constraint
	Config []ClaimConfiguration
}

type Request struct {
	Name string
	DeviceClassName string
	Selectors []Selector
	CountMode CountMode
	Count int64
}

type Selector struct {
	CEL *CELSelector
}

type Constraint struct {
	RequestNames []string
	MatchAttribute *string
}

type ClaimConfiguration struct {
	RequestNames []string
	Configuration
}
```

The place where this breaks down, I think, is when pod wants to xref a specific
request.  Would it make sense for pod to know that the request is a device?  Or
would it make sense to require that all requests (device, gizmo, or otherwise)
have a unique name within the claim?

How confident are you that gizmos will even have the notion of a "request"?

Reply:
I think the "request" level makes sense, no matter what gets requested.

What if we go into the other direction and eliminate the "device" indirection also in requests?

That means `DeviceRequest` gets folded into `Request`:
```
requests:
- deviceClassName: "dra.example.com"
  selectors:
  - cel:
      expression: device.attributes["dra.example.com"].driverVersion.isGreaterThan("1.0.0")
```

All of those fields are optional already, except for `deviceClassName`. If we need a different way of requesting `gizmos`, such a request could leave all of the current fields (including `deviceClassName`) unset and set something else instead.

```
requests:
- gizmoCategory: "magic" # intentionally not referencing a class object here, just for the sake of the argument
  # some more fields where I don't know what they might be...
```

Perhaps (probably?) `gizmos` even share some request fields with devices, like `CountMode`.

We wouldn't have the one-of-ness through the `Device *DeviceRequest` pointer anymore, but current clients would still fail safely because they don't get the expected `deviceClassName`. I'm still not a fan of making it required, but here this aspect of it comes handy :grin:

Reply:
In the `RequestAllocationResult` I had already removed the `Device` indirection because `Request` name, `Driver` name, and `Pool` name are generic. We then have `Device` name as the only field which is specific to devices: we need such a string to complete the `driver/pool/device` ID triplet.

If we add gizmos, that's fine? `Device` could become optional, with `Gizmo` as the alternative. Exactly one of them needs to be set. Making `Device` optional seems safe: current clients will get an empty string and fail when trying to use it until they get updated to also check the `Gizmo` alternative.

Reply:
> IOW, I vote "nay" on reducing explicitness.

Ack.

Your first proposal is the indirection embedded in each struct, which you said you "don't love" and I agree. Therefore I assume that with "I prefer this more" you mean the proposal at the end with `Devices []DeviceClaim`. That one looks good to me.

> I did not follow your point about status?

I was talking about:
```
type AllocationResult struct {
	Results []RequestAllocationResult
	...
}

type RequestAllocationResult struct {
	Request string
	Driver string
	Pool string
	Device string
}
```

With the new proposal, that becomes:
```yaml
type AllocationResult struct {
	Devices []DeviceAllocationResult
	...
}

type DeviceAllocationResult struct {
	Request string
	Driver string
	Pool string
	Device string
}
```

That's nice and clean, it just has one problem: kubelet only knows how to deal with "devices". Once we add "gizmos", kubelet must be updated, although it does the exact same thing for gizmos and devices. But I think that's acceptable.

Reply:
Some notes as I am going through implementing this...

- `s/Devices []DeviceClaim/Devices DeviceClaim/` in `ResourceClaimSpec` (it contains slices, but is not one itself).
- `ResourceClaimSpec` is not a one-of. We want to allow requesting one device and one gizmo (simpler than two claims, might want to define cross-type-constraints). Because it's not a one-of, adding gizmo support will be harder (ignored by old clients).
- Let's keep the request name unique across all requests, whether they are for devices or gizmos. That means we can keep using pod.spec.containers[].resources.claims[].request to identify exactly which request is meant, without having to add `deviceRequest`.
- Speaking of pod spec, I need to s/RequestName/Request/` for consistency with the "Name suffix only for resource types", and I can drop the pointer.
- If we want to support the `oneOf` "give me device A, otherwise B" use case later, we have to keep the `DeviceRequestDetail` indirection. We currently use `device` as field name there and I can't think of some better name (but see below).
- Limits on the size of slices need to be defined such that the *combined* slice sizes across the claim are considered, like it was done for attributes and capacities. I might not define it like that right now because the API does not have gizmos, but that is the intent. Otherwise adding gizmos would either a) increase the overall maximum claim size or b) imply reducing the limits for devices (a no-go).
- Thanks to workspace mode, `gopls` handles renames across *all* packages. Yeah!

With this new structure, we now have:
```yaml
apiVersion: resource.k8s.io/v1alpha3
kind: ResourceClaim
metadata:
  name: example
  namespace: default
spec:
  devices:
    requests:
    - name: req-0
      device:
        deviceClassName: example
```

Can we eliminate the inner `device`? Something like this:
```go
type DeviceRequest struct {
	Name string

         DeviceRequestDetail // inline

         // Future:
         // OneOf *OneOfDevices
}

type DeviceRequestDetail struct {
	// +required
	DeviceClassName string

        ....
}
```

It's hacky because `DeviceRequestDetail` no longer has cleanly separate alternatives. But perhaps they don't have to be alternatives?!

An empty OneOf makes no sense. We could define it as an optional field that adds alternative `DeviceRequestDetail` instances:
```go
type DeviceRequest struct {
	Name string
         DeviceRequestDetail // inline

         // Alternatives defines device requests that are considered
         // if and only if allocating the main one is not possible.
         // 
         // +optional
         Alternatives []DeviceRequestDetail
}
```

This leads to:
```yaml
apiVersion: resource.k8s.io/v1alpha3
kind: ResourceClaim
metadata:
  name: example
  namespace: default
spec:
  devices:
    requests:
    - name: req-0
      deviceClassName: dra.example.com
      alternatives:
       - deviceClassName: dra2.example.com
       - deviceClassName: dra3.example.com
```

However, `kubectl get -o yaml` will not use that ordering. With "or" instead of "alternatives" it's better, but not perfect: "selectors" still comes after it.

Reply:
I have pushed my implementation of this, with inner `device`.

Reply:
> However, kubectl get -o yaml will not use that ordering. With "or" instead of "alternatives" it's better, but not perfect: "selectors" still comes after it.

`XOr`? It's technically correct ("give me *either* device A *or* devive B") and it sorts last. But I have mixed feelings about it.

Reply:
Yet another alternative is to have a discriminator in `DeviceRequest`. I don't like that option because it becomes user-visible now for no obvious reason, but let me show what it looks like anyway:
```
requests:
- name: one-device-example
  requestMode: Simple # I have problems coming up with a good name...
  deviceClassName: dra.example.com
- name: one-of-example
  requestMode: OneOf
  OneOf:
  - deviceClassName: dra.example.com
  - deviceClassName: dra2.example.com
  - deviceClassName: dra3.example.com
```

My preference for 1.31 is to inline the struct and not add a discriminator, with clear guidance to clients that they *must* check for the `deviceClassName`. If we decide to evolve the API in such a way that the inlined struct becomes mutually exclusive with some other, future field, then we can do so safely because clients will react properly to the empty struct.

Reply:
Yet another option: calling it "spec" instead of "device".

```
requests:
- name: one-device-example
  spec:
    deviceClassName: dra.example.com
- name: one-of-example
  oneOf:
  - deviceClassName: dra.example.com
  - deviceClassName: dra2.example.com
  - deviceClassName: dra3.example.com
```

Reply:
> My preference for 1.31 is to inline the struct and not add a discriminator, with clear guidance to clients that they must check for the deviceClassName

Here's what that would look like:
```go
// DeviceRequest is a request for devices required for a claim.
// This is typically a request for a single resource like a device, but can
// also ask for several identical devices.
//
// An empty embedded DeviceRequestDetail is invalid at the moment. Clients must
// check this and refuse to handle such a request because it indicates that the
// DeviceRequest was extended in a way that the client doesn't understand yet.
type DeviceRequest struct {
	// Name can be used to reference this request in a pod.spec.containers[].resources.claims
	// entry and in a constraint of the claim.
	//
	// Must be a DNS label.
	//
	// +required
	Name string `json:"name" protobuf:"bytes,1,name=name"`

	DeviceRequestDetail `json:",inline" protobuf:"bytes,2,name=deviceRequestDetail"`
}

// DeviceRequestDetail is embedded in DeviceRequest and defines one request.
type DeviceRequestDetail struct {
	// DeviceClassName references a specific DeviceClass, which can define
	// additional configuration and selectors to be inherited by this
	// request.
	//
	// A class is required. Which classes are available depends on the cluster.
	//
	// Administrators may use this to restrict which devices may get
	// requested by only installing classes with selectors for permitted
	// devices. If users are free to request anything without restrictions,
	// then administrators can create an empty DeviceClass for users
	// to reference.
	//
	// +required
	DeviceClassName string `json:"deviceClassName" protobuf:"bytes,1,name=deviceClassName"`

...
```

Reply:
> kubelet only knows how to deal with "devices". Once we add "gizmos", kubelet must be updated, although it does the exact same thing for gizmos and devices. But I think that's acceptable.

and

> Because it's not a one-of, adding gizmo support will be harder (ignored by old clients).

I am STILL skeptical of gizmos' existence, in general, but yes.  If we add a field that is not part of an explicit one-of (which is easily detected, but still awkward to handle), then you have an unknown-fields-get-dropped miscomprehension wrt version skew.  This is true for any two components, but we are never far behind in the case of control plane components.  Kubelet can be N-3 (https://kubernetes.io/releases/version-skew-policy/#kubelet) which means that, for every new field we have to decide: Is it OK for Kubelet to process this pod without knowing that it dropped a field, or do we have to wait in alpha until N+3 (over a year) before allowing people to use it?  Or we need to find some other way to manage skew.

In the case of a devices+gizmos claim, it's either OK for kubelet to not enact the gizmos part (silently, btw - I am not sure there's even a way to know it dropped fields), or we add gizmos and wait wait wait, or we structure the API here as a repeated one-of, such that kubelet has clear indication that it should not proceed.  That doesn't help with the fact that the pod and claim were assigned to a node which can't handle gizmos. Maybe we need to have nodes self-report more information: https://github.com/kubernetes/enhancements/pull/4728

> Let's keep the request name unique across all requests, whether they are for devices or gizmos.

SGTM

> Limits on the size of slices need to be defined such that the combined slice sizes across the claim are considered

Yep.  This could be seen as another argument for not having devices and gizmos in the same claim, but since gizmos don't exist it's just so abstract to think about.

WRT DeviceRequestDetail...

That struct is:

```
{
	DeviceClassName string // required
	Selectors []DeviceSelector // optional
	CountMode DeviceCountMode // optional w/ default
	Count int64 // optional
	AdminAccess bool // optional w/ default
}
```

Today we say that `DeviceClassName` is required, but I know you still think of it as candidate to become optional.  As long as it is required, we CAN key off of it and detail this as: "if DeviceClassName is non-empty, the request is for a specific device shape, otherwise this request is a different mode and one of X, Y, Z must be set".  That's not horrible, IMO.

Describing it as `Alternatives` makes further evolution less obvious -- can we imagine something besides "this specific flavor" XOR "one of N flavors" that we could want?  If no, then maybe alternatives is not terrible?  If yes, I would avoid making this struct require a paragraph of natural language to explain.

> However, kubectl get -o yaml will not use that ordering

We have this problem in spades.  The way we render JSON/YAML is strictly correct but not super human friendly.  I would not advocate playing games with field names just to get a different sort order.  https://github.com/kubernetes/kubernetes/issues/125525

> another alternative is to have a discriminator in DeviceRequest

We can add a discriminator if/when we add a second mode, and it can (must) default to the historically implied value (e.g. "SpecificFlavor").  So a client who sends:

```
requests:
- name: one-device-example
  deviceClassName: dra.example.com
```

...will still work - the unstated value of `requestMode` is "Simple" (whatever we call it).  Anyone who wants to do something more complex would send:

```
requests:
- name: one-of-example
  requestMode: OneOf
  OneOf:
  - deviceClassName: dra.example.com
  - deviceClassName: dra2.example.com
  - deviceClassName: dra3.example.com
```

That said, if we have ANY OTHER field which is required (e.g. class), then this is strictly redundant (though clearer!)

The argument for including the mode from day one is to encourage clients to be aware that there may be alternative modes.  What I mean is what you called out elsewhere - a potential one-of group with one value (`deviceClassName`) where that value is documented as `optional` but is effectively required because of the one-of.

I think I am fine to document that `deviceClassName` is strictly optional, but effectively required until/unless other request modes are added, at which time it would be truly optional.

Aside: We do not have a formal definition of a `+oneOf=...` tag, but we could pioneer it, and start adding it in places we think will be treated that way.  Because it doesn't MEAN anything, we can afford to get it wrong.  Something like `// +oneOf=<name of a discriminator field, even if non-existent>`.  E.g. `+oneOf=RequestMode`.

> Yet another option: calling it "spec" instead of "device".

NAK please :)

Reply:
> I would not advocate playing games with field names just to get a different sort order.

Agreed. The better solution is to make a DeviceRequest either "simple" or a oneOf list of simple requests.

> The argument for including the mode from day one is to encourage clients to be aware that there may be alternative modes. 

So should we we do that? If yes, then we need to come up with names for 1.31. `requestMode: simple`? Not sure. I think I prefer to punt this till later and just use the wording in my [last comment above](https://github.com/kubernetes/kubernetes/pull/125488#discussion_r1667401243):

```
// An empty embedded DeviceRequestDetail is invalid at the moment. Clients must
// check this and refuse to handle such a request because it indicates that the
// DeviceRequest was extended in a way that the client doesn't understand yet.
```

> We do not have a formal definition of a +oneOf=... tag, but we could pioneer it, and start adding it in places we think will be treated that way.

I can add that in my next pass through the types.go.

Reply:
I went ahead and removed the "device" indirection *without* adding a discriminator.

@thockin : I don't mind adding the discriminator if you prefer that. If so, please let me know which name (`requestMode`?) and constant (`Simple`?) you prefer.

Reply:
I pushed the new commits just now.